### PR TITLE
Alternate fix for "Transformation breaks on PHP 7.x"

### PR DESCRIPTION
**Pull request by reedy**
Fix from https://phabricator.wikimedia.org/T215746#4944830

Ping https://github.com/cssjanus/php-cssjanus/issues/14 and https://github.com/cssjanus/php-cssjanus/pull/13

### CODE REVIEW

**Comment by Krinkle** [comment]
Some of these changes seem bug fixes for escaping. Others seem slightly more nominal in behaviour (towards the bottom of the diff). For those, we should make sure to update the canonical JS implementation as well, preferably first so that we avoid diverging.

**Comment by reedy** [comment]
#22 was an example of https://github.com/cssjanus/php-cssjanus/pull/13#issuecomment-355574605 in practice... (it failed)

File: src/CSSJanus.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ class CSSJanus {
 	private static $patterns = array(
 		'tmpToken' => '`TMP`',
 		'nonAscii' => '[\200-\377]',
-		'unicode' => '(?:(?:\\[0-9a-f]{1,6})(?:\r\n|\s)?)',
+		'unicode' => '(?:(?:\\\\[0-9a-f]{1,6})(?:\r\n|\s)?)',
```
Comment (reedy) [inline src/CSSJanus.php]: Already upstream https://github.com/cssjanus/cssjanus/blob/f5b34a1/src/cssjanus.js#L101

File: src/CSSJanus.php [hunk 2]
```diff
@@ -92,17 +92,17 @@ private static function buildPatterns() {
 
 		// @codingStandardsIgnoreStart Generic.Files.LineLength.TooLong
 		$patterns =& self::$patterns;
-		$patterns['escape'] = "(?:{$patterns['unicode']}|\\[^\r\n\f0-9a-f])";
+		$patterns['escape'] = "(?:{$patterns['unicode']}|\\\\[^\\r\\n\\f0-9a-f])";
```
Comment (reedy) [inline src/CSSJanus.php]: Already upstream https://github.com/cssjanus/cssjanus/blob/f5b34a1/src/cssjanus.js#L101

File: src/CSSJanus.php [hunk 3]
```diff
@@ -92,17 +92,17 @@ private static function buildPatterns() {
...
 		$patterns['nmchar'] = "(?:[_a-z0-9-]|{$patterns['nonAscii']}|{$patterns['escape']})";
 		$patterns['ident'] = "-?{$patterns['nmstart']}{$patterns['nmchar']}*";
 		$patterns['quantity'] = "{$patterns['num']}(?:\s*{$patterns['unit']}|{$patterns['ident']})?";
 		$patterns['possibly_negative_quantity'] = "((?:-?{$patterns['quantity']})|(?:inherit|auto))";
 		$patterns['color'] = "(#?{$patterns['nmchar']}+|(?:rgba?|hsla?)\([ \d.,%-]+\))";
 		$patterns['url_chars'] = "(?:{$patterns['url_special_chars']}|{$patterns['nonAscii']}|{$patterns['escape']})*";
-		$patterns['lookahead_not_open_brace'] = "(?!({$patterns['nmchar']}|\r?\n|\s|#|\:|\.|\,|\+|>|\(|\)|\[|\]|=|\*=|~=|\^=|'[^']*'])*?{)";
-		$patterns['lookahead_not_closing_paren'] = "(?!{$patterns['url_chars']}?{$patterns['valid_after_uri_chars']}\))";
-		$patterns['lookahead_for_closing_paren'] = "(?={$patterns['url_chars']}?{$patterns['valid_after_uri_chars']}\))";
+		$patterns['lookahead_not_open_brace'] = "(?!({$patterns['nmchar']}|\\r?\\n|\s|#|\:|\.|\,|\+|>|\(|\)|\[|\]|=|\*=|~=|\^=|'[^']*'])*+{)";
```
Comment (reedy) [inline src/CSSJanus.php]: Additional \ are in Already upstream https://github.com/cssjanus/cssjanus/blob/f5b34a1/src/cssjanus.js#L101

**Comment by reedy** [comment]
> Some of these changes seem bug fixes for escaping. Others seem slightly more nominal in behaviour (towards the bottom of the diff). For those, we should make sure to update the canonical JS implementation as well, preferably first so that we avoid diverging.

See attached comments. Escaping fixes are already upstream

https://github.com/cssjanus/cssjanus/pull/73 brings the 3 changes of `?` for `+`

File: src/CSSJanus.php (hunk 1 above)
Comment (reedy) [reply src/CSSJanus.php]: https://github.com/cssjanus/cssjanus/commit/d0d6a1ce0c65b1760236e5e91d932097293ffa2b (in 1.3.1)

File: src/CSSJanus.php (hunk 2 above)
Comment (reedy) [reply src/CSSJanus.php]: Always been upstream, must've been eaten in translation https://github.com/cssjanus/cssjanus/commit/b38c7e1bbe8662c01713643021cae6c253997a64

**Comment by reedy** [comment]
https://github.com/cssjanus/php-cssjanus/pull/25 splits off the escaping changes, syncing it with upstream (some of which were there already, some have been fixed in more recent commits)

Might help making this commit cleaner (can be rebased to remove the escaping changes)

File: src/CSSJanus.php [hunk 4]
```diff
@@ -92,17 +92,17 @@ private static function buildPatterns() {
 
 		// @codingStandardsIgnoreStart Generic.Files.LineLength.TooLong
 		$patterns =& self::$patterns;
-		$patterns['escape'] = "(?:{$patterns['unicode']}|\\\\[^\\r\\n\f0-9a-f])";
+		$patterns['escape'] = "(?:{$patterns['unicode']}|\\\\[^\\r\\n\\f0-9a-f])";
```
Comment (reedy) [inline src/CSSJanus.php]: Looks like the \\f was already upstream. Should've gone in https://github.com/cssjanus/php-cssjanus/pull/25 but it was hiding in plain sight :)

Comment (reedy) [reply src/CSSJanus.php]: https://github.com/cssjanus/php-cssjanus/pull/26

**Comment by reedy** [comment]
Only thing outstanding here is now (I think) is a test case, which is done upstream in https://github.com/cssjanus/cssjanus/commit/1a5a1c7c63a0d93e95773b618564ba2cc89540fe

Just needs a new release tagging, this PR updating to use it, which then fixes #20

**Pull request merged**